### PR TITLE
Resource gatherer refactoring

### PR DESCRIPTION
**Pull request by gmarek**
Ref. #21313

### CODE REVIEW

File: test/e2e/resource_usage_gatherer.go [hunk 1]
```diff
@@ -75,43 +70,193 @@ func (s *ResourceUsageSummary) PrintJSON() string {
...
 		}
+	case <-w.stopCh:
+		return
+	}
+}
+
+func (g *containerResourceGatherer) getKubeSystemContainersResourceUsage(c *client.Client) error {
+	errors := make([]error, 0)
+	delay := resourceDataGatheringPeriod / time.Duration(len(g.workers))
+	go func() {
```
Comment (wojtek-t) [inline test/e2e/resource_usage_gatherer.go]: please remove this "go" - you're spawning only go routines here

File: test/e2e/resource_usage_gatherer.go [hunk 2]
```diff
@@ -75,43 +70,193 @@ func (s *ResourceUsageSummary) PrintJSON() string {
...
+		for i := range g.workers {
+			go g.workers[i].gather(delay, errors)
```
Comment (wojtek-t) [inline test/e2e/resource_usage_gatherer.go]: Passing "errors" here, makes it thread-unsafe.
In fact generally it seems that at the end of the day you are ignoring those errors anyway.

So what i would suggest is NOT passing "errors" at all to "gather" method, and only log those errors where they appear.

File: test/e2e/resource_usage_gatherer.go [hunk 3]
```diff
@@ -75,43 +70,193 @@ func (s *ResourceUsageSummary) PrintJSON() string {
...
+		client:               c,
+		stopCh:               make(chan struct{}),
+		containerIDToNameMap: make(map[string]string),
+		containerIDs:         make([]string, 0),
+	}
+
+	pods, err := c.Pods("kube-system").List(api.ListOptions{})
+	if err != nil {
+		Logf("Error while listing Pods: %v", err)
+		return nil
```
Comment (wojtek-t) [inline test/e2e/resource_usage_gatherer.go]: return err

Comment (wojtek-t) [inline test/e2e/resource_usage_gatherer.go]: Actually no - so probably NewRUG() should also return an error.

File: test/e2e/resource_usage_gatherer.go [hunk 4]
```diff
@@ -75,43 +70,193 @@ func (s *ResourceUsageSummary) PrintJSON() string {
...
+	}
+	for _, pod := range pods.Items {
+		for _, container := range pod.Status.ContainerStatuses {
+			containerID := strings.TrimPrefix(container.ContainerID, "docker:/")
+			g.containerIDToNameMap[containerID] = pod.Name + "/" + container.Name
+			g.containerIDs = append(g.containerIDs, containerID)
+		}
+	}
+	var nodeList *api.NodeList
+	if wait.PollImmediate(poll, 10*time.Second, func() (bool, error) {
```
Comment (wojtek-t) [inline test/e2e/resource_usage_gatherer.go]: Did you see this being a problem? Why not simply nodes, err L- c.Nodes().List(api.ListOptions{}) ?

File: test/e2e/resource_usage_gatherer.go [hunk 5]
```diff
@@ -75,43 +70,193 @@ func (s *ResourceUsageSummary) PrintJSON() string {
...
+			containerIDToNameMap: g.containerIDToNameMap,
+			containerIDs:         g.containerIDs,
+			stopCh:               g.stopCh,
+		})
+	}
+	return &g
+}
+
+func (g *containerResourceGatherer) startGatheringData() {
+	go g.getKubeSystemContainersResourceUsage(g.client)
```
Comment (wojtek-t) [inline test/e2e/resource_usage_gatherer.go]: maybe simply paste the content of "getKubeSystemContainersRU" here and say that startGatheringData() is blocking and run it with go

File: test/e2e/resource_usage_gatherer.go (hunk 1 above)
Comment (gmarek) [inline test/e2e/resource_usage_gatherer.go]: Yes, sorry - it's a leftover from previous version.

File: test/e2e/resource_usage_gatherer.go (hunk 2 above)
Comment (gmarek) [inline test/e2e/resource_usage_gatherer.go]: Done.

File: test/e2e/resource_usage_gatherer.go (hunk 4 above)
Comment (gmarek) [inline test/e2e/resource_usage_gatherer.go]: Done.

File: test/e2e/resource_usage_gatherer.go (hunk 5 above)
Comment (gmarek) [inline test/e2e/resource_usage_gatherer.go]: Done.

File: test/e2e/framework.go [hunk 6]
```diff
@@ -116,7 +116,11 @@ func (f *Framework) beforeEach() {
 	}
 
 	if testContext.GatherKubeSystemResourceUsageData {
-		f.gatherer.startGatheringData(c, resourceDataGatheringPeriodSeconds*time.Second)
+		f.gatherer, err = NewResourceUsageGatherer(c)
+		if err != nil {
+			Logf("Error while creating NewResourceUsageGatherer: %v", err)
+		}
+		go f.gatherer.startGatheringData()
```
Comment (wojtek-t) [inline test/e2e/framework.go]: check if f.gatherer is not nil

File: test/e2e/resource_usage_gatherer.go [hunk 7]
```diff
@@ -75,43 +69,186 @@ func (s *ResourceUsageSummary) PrintJSON() string {
...
+	}
+}
+
+func (g *containerResourceGatherer) getKubeSystemContainersResourceUsage(c *client.Client) error {
+	delay := resourceDataGatheringPeriod / time.Duration(len(g.workers))
+	for i := range g.workers {
+		go g.workers[i].gather(delay)
+	}
+	g.workerWg.Wait()
+
```
Comment (wojtek-t) [inline test/e2e/resource_usage_gatherer.go]: remove empty line

File: test/e2e/resource_usage_gatherer.go [hunk 8]
```diff
@@ -75,43 +69,186 @@ func (s *ResourceUsageSummary) PrintJSON() string {
...
+	return nil
```
Comment (wojtek-t) [inline test/e2e/resource_usage_gatherer.go]: Does it make sense to return anything here? Maybe return possibility to return error from this function?

**Comment by wojtek-t** [comment]
Just few more minor things

File: test/e2e/resource_usage_gatherer.go (hunk 7 above)
Comment (gmarek) [inline test/e2e/resource_usage_gatherer.go]: Done.

File: test/e2e/resource_usage_gatherer.go (hunk 8 above)
Comment (gmarek) [inline test/e2e/resource_usage_gatherer.go]: Done.

File: test/e2e/framework.go (hunk 6 above)
Comment (gmarek) [inline test/e2e/framework.go]: Done.

**Comment by wojtek-t** [comment]
LGTM

**Pull request merged**